src/android/SecureStorage.java

```diff
@@ -66,192 +142,334 @@ public boolean execute(String action, CordovaArgs args, final CallbackContext ca
-            cordova.getThreadPool().execute(new Runnable() {
-                public void run() {
-                    try {
-                        byte[] decrypted = RSA.decrypt(decryptMe, service2alias(service));
-                        callbackContext.success(new String(decrypted));
-                    } catch (Exception e) {
-                        Log.e(TAG, "Decrypt (RSA) failed :", e);
-                        callbackContext.error(e.getMessage());
-                    }
-                }
-            });
-            return true;
+
+        if (!isDeviceSecure()) {
+            // Lock screen that requires authentication is not defined
+            Log.e(TAG, MSG_DEVICE_NOT_SECURE);
+            callbackContext.error(MSG_DEVICE_NOT_SECURE);
+
+        } else if (!RSA.isEntryAvailable(alias)) {
+            // Key for alias does not exist
+            handleLockScreen(IntentRequestType.INIT, service, callbackContext);
+
+        } else {
+            // No actions are required to init correctly
+            initSuccess(callbackContext);
         }
+        return true;
+    }
 
-        // Encrypt a message using the key with an alias associated with the store name
-        if ("encrypt_rsa".equals(action)) {
-            Log.v(TAG, "Called encrypt_rsa action");
-            final String service = args.getString(0);
-            final String encryptMe = args.getString(1);
-            cordova.getThreadPool().execute(new Runnable() {
-                public void run() {
-                    try {
-                        byte[] encrypted = RSA.encrypt(encryptMe.getBytes(), service2alias(service));
-                        callbackContext.success(Base64.encodeToString(encrypted, Base64.DEFAULT));
-                    } catch (Exception e) {
-                        Log.e(TAG, "Encrypt (RSA) failed :", e);
-                        callbackContext.error(e.getMessage());
-                    }
-                }
-            });
-            return true;
+    private boolean checkForSecurityMigration() {
+        Context ctx = getContext();
+        SharedPreferences preferences = ctx.getSharedPreferences(ctx.getPackageName() + "_SM",0);
+        String isMigrated = preferences.getString(MIGRATED_FOR_SECURITY, "");
+        //check OS then check if keys exist and migration was done
+        if(Build.VERSION.SDK_INT > Build.VERSION_CODES.M){
+
+            int size = initializePreferences();
+            Log.e(TAG, "TARGET OS");
+            if(isMigrated.equals("TRUE")){
+
+                Log.e(TAG, "MIGRATION ALREADY DONE");
+                return false;
+            }
+            //the target case of migration
+            else if(size > 0){
+
+                Log.e(TAG, "TARGET CASE FOR MIGRATION");
+                return true;
+            }
+            //new use, meaning we should put the tag in as to not trigger a unwanted migration
+            else{
+
+                Log.e(TAG, "NEW USE");
```

Comment:
Please make this entry more clear about what it means, since customers are able to view the logs.